### PR TITLE
[DeepSeek-V4-Flash][SM120] FlashInfer MXFP4 path with custom SM120 FlashMLA

### PR DESCRIPTION
**Pull request by ambientlight**
## Motivation

Native reference path for SM120 for dsv4-flash with:

1. **FlashInfer native MXFP4 W4A4 fused MoE:** https://github.com/flashinfer-ai/flashinfer/pull/3541
2. custom [sparse_decode_kernel.cuh](https://github.com/ambientlight/deepseek-v4-flash-sm120/blob/feat/hmma-tensor-core-sparse-decode/csrc/sm120/decode/sparse_decode_kernel.cuh) + [sparse_prefill_kernel.cuh](https://github.com/ambientlight/deepseek-v4-flash-sm120/blob/feat/hmma-tensor-core-sparse-decode/csrc/sm120/prefill/sparse_prefill_kernel.cuh) HMMA kernels from [deepseek-v4-flash-sm120](https://github.com/ambientlight/deepseek-v4-flash-sm120) as a drop-in replacement to DSv4 stock FlashMLA kernels unavailable for SM120

Not sure if there is way to check this in but putting this out as a practical reference for 4x RTX6000 pro max-q blackwell box. Reference setup per [DEPLOY-MXFP4-W4A4-DEEPSEEK-V4-FLASH-SM120.md](https://github.com/ambientlight/rtx-pro-6000-bench/blob/main/docs/DEPLOY-MXFP4-W4A4-DEEPSEEK-V4-FLASH-SM120.md) (assembled via hillclimbling with 4.8 opus max against [OxSero/deepseek-v4-flash-sm120 fork](https://github.com/ambientlight/deepseek-v4-flash-sm120))

## Modifications

1. MXFP4 W4A4 fused MoE: loads the checkpoint's MXFP4 experts as-is and runs the fused SwiGLU GEMM through FlashInfer's SM120 CuTe-DSL path ([flashinfer-ai#3541](https://github.com/flashinfer-ai/flashinfer/pull/3541)).
2. Selectable SM120 sparse decode: `SGLANG_SM120_SPARSE_DECODE` toggle (hmma, triton, torch; default triton). hmma is our custom [sparse_decode_kernel.cuh](https://github.com/ambientlight/deepseek-v4-flash-sm120/blob/feat/hmma-tensor-core-sparse-decode/csrc/sm120/decode/sparse_decode_kernel.cuh). If the custom kernel package isn't installed, hmma falls back to triton
3. Selectable SM120 sparse prefill: same idea for [sparse_prefill_kernel.cuh](https://github.com/ambientlight/deepseek-v4-flash-sm120/blob/feat/hmma-tensor-core-sparse-decode/csrc/sm120/prefill/sparse_prefill_kernel.cuh): a `SGLANG_SM120_SPARSE_PREFILL` toggle (hmma, sglkernel, torch). 
4. Add a split-KV variant d1c590ffc5f461ac2491bcc59a47dd282dfc4b7b of the SM120 tilelang paged-MQA-logits indexer kernel that shards each row's KV scan across (batch, num_splits) CTAs allowing to push long-context decode at `256K / 512K / 1M` from `20.7 / 12.4 / 6.9` to `27.4 / 17.7 / 10.4` tok/s

Both selectors share a small `is_deepseek_v4_kernel_available()` probe that decides whether hmma is reachable.

**also:** SM120 bugfix https://github.com/sgl-project/sglang/commit/e1ca8d9672d8ef2efe7b97c21d572e702054ee98 routes the tilelang FP8 paged-MQA-logits indexer to dsv4 instead of dsa kernel, fixing an illegal memory access under CUDA-graph capture

## Accuracy Tests

SWE-Bench Verified scores 76% on mini-swe-agent 2.4.2 harness per https://github.com/ambientlight/rtx-pro-6000-bench/tree/main/evals/swebench-verified/deepseek-v4-flash-2026-06-20

```
python3 -m sglang.test.few_shot_gsm8k --num-questions 200
# Accuracy: 0.980   Invalid: 0.000
```

The HMMA kernels are checked [deepseek-v4-flash-sm120/tests](https://github.com/ambientlight/deepseek-v4-flash-sm120/tree/feat/hmma-tensor-core-sparse-decode/tests) against a torch port of `ref_sparse_attn_fwd`:

```
python -m pytest tests/test_sparse_decode.py tests/test_sparse_prefill.py -v -p no:cacheprovider

shape (s_q,h_q,topk)           out_cosine  max|out-ref|
(64,64,64)                      0.9999976      1.95e-03
(1024,128,256)                  0.9999976      9.77e-04
(64,128,2048)                   0.9999977      2.44e-04
(1024,64,256)                   0.9999977      9.77e-04
```

## Speed Tests and Profiling (4x RTX6000 pro blackwell max-q)

See [DEPLOY-MXFP4-W4A4-DEEPSEEK-V4-FLASH-SM120.md#performance](https://github.com/ambientlight/rtx-pro-6000-bench/blob/main/docs/DEPLOY-MXFP4-W4A4-DEEPSEEK-V4-FLASH-SM120.md#performance)

[bench/deepseek-v4-flash_W300_TP4_sglang](https://github.com/ambientlight/rtx-pro-6000-bench/tree/main/bench/deepseek-v4-flash_W300_TP4_sglang): single-stream decode scales to the **full 1M context**: `64K / 128K / 256K / 512K / 1M` token prompts sustain `46 / 38 / 27 / 18 / 10 tok/s`, the 1M prompt peaking at 95% VRAM [bench_sweep_single.chunk8192.log](https://github.com/ambientlight/rtx-pro-6000-bench/blob/main/bench/deepseek-v4-flash_W300_TP4_sglang/bench_sweep_single.chunk8192.log), concurrency sweep at W300/TP4 [bench_sweep.log](https://github.com/ambientlight/rtx-pro-6000-bench/blob/main/bench/deepseek-v4-flash_W300_TP4_sglang/bench_sweep.log): **8576/8576 requests, 0 failed over ~27 h continuous load**, best sustained 756 tok/s @ 2K (c64)

## Checklist

- [x] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [x] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [x] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

## Review and Merge Process

1. Ping Merge Oncalls to start the process. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
2. Get approvals from [CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
3. Trigger CI tests with [comments](https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) or contact authorized users to do so.
   - Common commands include `/tag-and-rerun-ci`, `/tag-run-ci-label`, `/rerun-failed-ci`
4. After green CI and required approvals, ask Merge Oncalls or people with Write permission to merge the PR.

---
### CI States

Latest PR Test (Base): :x: [Run #27441191956](https://github.com/sgl-project/sglang/actions/runs/27441191956)
Latest PR Test (Extra): :x: [Run #27441191771](https://github.com/sgl-project/sglang/actions/runs/27441191771)

### CODE REVIEW

File: python/sglang/srt/layers/attention/flash_mla_sparse_prefill_sm120.py [hunk 1]
```diff
@@ -0,0 +1,218 @@
...
+        P[invalid.unsqueeze(1).broadcast_to(P.shape)] = float("-inf")
+
+        orig_lse = torch.logsumexp(P, dim=-1)  # [n, h_q]
+        mx = P.max(dim=-1).values  # [n, h_q]
+
+        if sink is not None:
+            lse_for_o = torch.logsumexp(
+                torch.stack([orig_lse, sink.broadcast_to(n, h_q)], dim=0),
+                dim=0,
+            )
```
[the remark this answers is not in the document]
Comment (ambientlight) [reply python/sglang/srt/layers/attention/flash_mla_sparse_prefill_sm120.py:152]: `sink.broadcast_to(n, h_q)` is valid
